### PR TITLE
Allow writing with subtree owners in partial set cases (with isPartialSet = true)

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Allow writing rules / functions / owners with subtree owners in partial set cases (with isPartialSet = true)
- Do not return empty error_messages

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/779

### CODE REVIEW

**Comment by liayoo** [comment]
Could you help me understand the use case of partial sets? Like if we're going to allow users to specify is_partial_set, why can't we just allow it always? (without specifying is_partial_set option)

**Comment by platfowner** [comment]
> Could you help me understand the use case of partial sets? Like if we're going to allow users to specify is_partial_set, why can't we just allow it always? (without specifying is_partial_set option)

Thanks for asking. is_partial_set options is used for only eval APIs (ain_evalOwner or /eval_owner) APIs.
For set operations (SET_RULE, SET_FUNCTION, SET_OWNER), it's auto-detected by apply functions (applyRuleChange(), applyFunctionChange(), applyOwnerChange()).
For now this kind of auto-detection is impossible for the eval APIs as users are not allowed to pass the actual rule / function / owner config.
Let me check if we can allow such values as a parameter after this PR.

File: common/common-util.js [hunk 1]
```diff
@@ -255,6 +255,9 @@ class CommonUtil {
     if (args.is_global !== undefined) {
       options.isGlobal = CommonUtil.toBool(args.is_global);
     }
+    if (args.is_partial_set !== undefined) {
```
Comment (liayoo) [inline common/common-util.js:258]: as discussed offline, how about `is_merge`?

**Comment by liayoo** [comment]
> > Could you help me understand the use case of partial sets? Like if we're going to allow users to specify is_partial_set, why can't we just allow it always? (without specifying is_partial_set option)
> 
> Thanks for asking. is_partial_set options is used for only eval APIs (ain_evalOwner or /eval_owner) APIs. For set operations (SET_RULE, SET_FUNCTION, SET_OWNER), it's auto-detected by apply functions (applyRuleChange(), applyFunctionChange(), applyOwnerChange()). For now this kind of auto-detection is impossible for the eval APIs as users are not allowed to pass the actual rule / function / owner config. Let me check if we can allow such values as a parameter after this PR. (Just added a todo)

I see! Thanks for the detailed explanation. A to-do sounds good for now~

File: common/common-util.js (hunk 1 above)
Comment (platfowner) [reply common/common-util.js]: Just updated, thanks!

**Comment by platfowner** [comment]
Comments addressed, PTAL~ @liayoo

File: db/state-util.js [hunk 2]
```diff
@@ -607,34 +607,53 @@ function hasConfigLabelOnly(stateTreeObj, configLabel) {
...
  * the current rule tree.
  * 
  * @param {Object} curRuleTree current rule tree (to be modified by this rule)
  * @param {Object} ruleChange rule change
  */
-// NOTE(platfowner): Partial set is applied only when the current rule tree has
+// NOTE(platfowner): Config merge is applied only when the current rule tree has
 // .rule property and the rule change has .rule property as the only property.
 function applyRuleChange(curRuleTree, ruleChange) {
+  // 1. Config overwriting case:
```
Comment (liayoo) [inline db/state-util.js:630]: thanks for the comments! 👍

File: db/state-util.js [hunk 3]
```diff
@@ -607,34 +607,53 @@ function hasConfigLabelOnly(stateTreeObj, configLabel) {
...
-    return CommonUtil.isDict(ruleChange) ?
+    const newRuleConfig = CommonUtil.isDict(ruleChange) ?
         JSON.parse(JSON.stringify(ruleChange)) : ruleChange;
+    return {
+      isMerge: false,
+      ruleConfig: newRuleConfig,
+    };
   }
-  const ruleChangeMap = CommonUtil.getJsObject(ruleChange, [PredefinedDbPaths.DOT_RULE]);
+  // 2. Config merge case:
```
Comment (liayoo) [inline db/state-util.js:640]: minor) how about no changes for this case & config merge case for the below?

File: db/state-util.js [hunk 4]
```diff
@@ -607,34 +607,53 @@ function hasConfigLabelOnly(stateTreeObj, configLabel) {
...
  * 
  * @param {Object} curRuleTree current rule tree (to be modified by this rule)
  * @param {Object} ruleChange rule change
  */
-// NOTE(platfowner): Partial set is applied only when the current rule tree has
+// NOTE(platfowner): Config merge is applied only when the current rule tree has
 // .rule property and the rule change has .rule property as the only property.
 function applyRuleChange(curRuleTree, ruleChange) {
+  // 1. Config overwriting case:
   if (!hasConfigLabel(curRuleTree, PredefinedDbPaths.DOT_RULE) ||
```
Comment (liayoo) [inline db/state-util.js:631]: I think this condition is covered with !hasConfigLabelOnly() check?

Comment (platfowner) [reply db/state-util.js:631]: Ah, this is checked for 'curRuleTree' and the other one is for 'ruleChange'.

File: db/state-util.js (hunk 3 above)
Comment (platfowner) [reply db/state-util.js]: just done, thanks!

File: db/state-util.js (hunk 4 above)
Comment (liayoo) [reply db/state-util.js:631]: Right, missed that. Thanks!

**Review by liayoo — APPROVED** [review]
LGTM, thanks!

**Comment by platfowner** [comment]
Thanks for the fast review, now merging..

**Pull request merged**